control-plane-pki-operator/certificatesigningcontroller/certificatesigningcontroller_test.go

```diff
@@ -0,0 +1,528 @@
+						{
+							Type:    ptr.To(certificatesv1.CertificateFailed),
+							Status:  ptr.To(corev1.ConditionTrue),
+							Reason:  ptr.To("SignerValidationFailure"),
+							Message: ptr.To("invalid"),
+						},
+					},
+				},
+			},
+			expectedValidationErr: true,
+		},
+		{
+			description: "valid csr",
+			name:        "test-csr",
+			getCSR: func(name string) (*certificatesv1.CertificateSigningRequest, error) {
+				if name != "test-csr" {
+					return nil, apierrors.NewNotFound(certificatesv1.SchemeGroupVersion.WithResource("certificatesigningrequests").GroupResource(), name)
+				}
+				return &certificatesv1.CertificateSigningRequest{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: name,
+					},
+					Spec: testCSRSpec(csrBuilder{
+						cn:         "test-client",
+						org:        []string{"system:masters"},
+						signerName: "hcp-name.hcp-namespace.hypershift.openshift.io/customer-break-glass",
+						usages:     []certificatesv1.KeyUsage{certificatesv1.UsageClientAuth},
+					}),
+					Status: certificatesv1.CertificateSigningRequestStatus{
+						Conditions: []certificatesv1.CertificateSigningRequestCondition{{
+							Type:   certificatesv1.CertificateApproved,
+							Status: corev1.ConditionTrue,
+						}},
+					},
+				}, nil
+			},
+			validator: certificates.Validator(&hypershiftv1beta1.HostedControlPlane{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "hcp-namespace",
+					Name:      "hcp-name",
+				},
+			}, certificates.CustomerBreakGlassSigner),
+			expectedCfg: &certificatesv1applyconfigurations.CertificateSigningRequestApplyConfiguration{
+				Status: &certificatesv1applyconfigurations.CertificateSigningRequestStatusApplyConfiguration{
+					Certificate: []uint8(`testdata`),
+				},
+			},
+		},
+	}
+
+	for _, testCase := range cases {
+		t.Run(testCase.description, func(t *testing.T) {
+			ca := certificateAuthority(t)
+
+			c := CertificateSigningController{
+				validator: testCase.validator,
+				getCSR:    testCase.getCSR,
+				getCurrentCABundleContent: func() (*authority.CertificateAuthority, error) {
+					return ca, nil
+				},
+				certTTL: 12 * time.Hour,
+			}
+
+			cfg, _, validationErr, err := c.processCertificateSigningRequest(testCase.name, fakeClock.Now)
+			if testCase.expectedErr && err == nil {
+				t.Errorf("expected an error but got none")
+			} else if !testCase.expectedErr && err != nil {
+				t.Errorf("expected no error but got: %v", err)
+			}
+			if testCase.expectedValidationErr && validationErr == nil {
+				t.Errorf("expected a validation error but got none")
+			} else if !testCase.expectedValidationErr && validationErr != nil {
+				t.Errorf("expected no validation error but got: %v", validationErr)
+			}
+
+			// signing the certificate necessarily uses cryptographic randomness, so we can't know
+			// what the output will be a priori
+			if testCase.expectedCfg != nil && testCase.expectedCfg.Status != nil && testCase.expectedCfg.Status.Certificate != nil &&
```

Comment:
This is unfortunate but I don't know of any way for us to avoid it. Essentially we end up asserting that if the test case expected *something* in the signed certificate field, there is some value  - any value - there.